File: src/com/colacelli/ircbot/plugins/loader/LoaderPlugin.java
package com.colacelli.ircbot.plugins.loader;

import com.colacelli.ircbot.IRCBot;
import com.colacelli.ircbot.Plugin;
import com.colacelli.ircbot.listeners.OnChannelCommandListener;
import com.colacelli.ircbot.plugins.access.IRCBotAccess;
import com.colacelli.ircbot.plugins.help.PluginHelp;
import com.colacelli.ircbot.plugins.help.PluginHelper;
import com.colacelli.irclib.connection.Connection;
import com.colacelli.irclib.messages.ChannelMessage;
import com.colacelli.irclib.messages.PrivateNoticeMessage;

import java.util.ArrayList;

public class LoaderPlugin implements Plugin {
    @Override
    public String getName() {
        return "LOADER";
    }

    @Override
    public void onLoad(IRCBot bot) {
        IRCBotAccess.getInstance().addListener(bot, IRCBotAccess.SUPER_ADMIN_LEVEL, new OnChannelCommandListener() {
            @Override
            public String channelCommand() {
                return ".plugin";
            }

            @Override
            public void onChannelCommand(Connection connection, ChannelMessage message, String command, String... args) {
                if (args != null) {
                    ArrayList<Plugin> botPlugins = bot.getPlugins();
                    PrivateNoticeMessage.Builder builder = new PrivateNoticeMessage.Builder();
                    builder
                            .setSender(connection.getUser())
                            .setReceiver(message.getSender());

                    if (args.length > 1) {
                        switch (args[0]) {
                            case "load":
                                for (int i = 0; i < botPlugins.size(); i++) {
                                    Plugin plugin = botPlugins.get(i);
                                    if (plugin.getName().toUpperCase().equals(args[1].toUpperCase())) {
                                        plugin.onLoad(bot);
                                        builder.setText(plugin.getName() + " loaded!");
                                        connection.send(builder.build());
                                    }
                                }
                                break;

                            case "unload":
                                for (int i = 0; i < botPlugins.size(); i++) {
                                    Plugin plugin = botPlugins.get(i);
                                    if (plugin.getName().toUpperCase().equals(args[1].toUpperCase())) {
                                        plugin.onUnload(bot);
                                        builder.setText(plugin.getName() + " unloaded!");
                                        connection.send(builder.build());
                                    }
                                }
                                break;
                        }
                    } else {
                        switch (args[0]) {
                            case "list":
                                botPlugins.forEach((plugin) -> {
                                    builder.setText(plugin.getName());
                                    connection.send(builder.build());
                                });
                                break;
                        }
                    }
                }
            }
        });

        PluginHelper.getInstance().addHelp(new PluginHelp(
                ".plugin list",
                IRCBotAccess.SUPER_ADMIN_LEVEL,
                "List all available plugins"));

        PluginHelper.getInstance().addHelp(new PluginHelp(
                ".plugin load",
                IRCBotAccess.SUPER_ADMIN_LEVEL,
                "Load a plugin",
                "plugin"));

        PluginHelper.getInstance().addHelp(new PluginHelp(
                ".plugin unload",
                IRCBotAccess.SUPER_ADMIN_LEVEL,
                "Unload a plugin",
                "plugin"));
    }

    @Override
    public void onUnload(IRCBot bot) {
        IRCBotAccess.getInstance().removeListener(bot, ".plugin");
        PluginHelper.getInstance().removeHelp(".plugin list");
        PluginHelper.getInstance().removeHelp(".plugin load");
        PluginHelper.getInstance().removeHelp(".plugin unload");
    }
}

File: src/com/colacelli/ircbot/plugins/loader/LoaderPlugin.kt
package com.colacelli.ircbot.plugins.loader

import com.colacelli.ircbot.IRCBot
import com.colacelli.ircbot.Plugin
import com.colacelli.ircbot.listeners.OnChannelCommandListener
import com.colacelli.ircbot.plugins.access.IRCBotAccess
import com.colacelli.ircbot.plugins.help.PluginHelp
import com.colacelli.ircbot.plugins.help.PluginHelper
import com.colacelli.irclib.connection.Connection
import com.colacelli.irclib.messages.ChannelMessage
import com.colacelli.irclib.messages.PrivateNoticeMessage

class LoaderPlugin : Plugin {
    override fun getName(): String {
        return "plugin_loader"
    }

    override fun onLoad(bot: IRCBot) {
        IRCBotAccess.instance.addListener(bot, IRCBotAccess.Level.ROOT, object : OnChannelCommandListener {
            override fun channelCommand(): String {
                return ".plugin"
            }

            override fun onChannelCommand(connection: Connection, message: ChannelMessage, command: String, args: Array<String>) {
                val message = PrivateNoticeMessage.Builder()
                        .setSender(connection.user)
                        .setReceiver(message.sender)

                if (args.size > 1) {
                    when (args[0]) {
                        "load" -> {
                            val plugin = args[1].toLowerCase()
                            bot.plugins.forEach {
                                if (it.getName().toLowerCase() == plugin) {
                                    it.onLoad(bot)

                                    message.setText("Plugin loaded!")
                                    connection.send(message.build())
                                }
                            }
                        }

                        "unload" -> {
                            val plugin = args[1].toLowerCase()
                            bot.plugins.forEach {
                                if (it.getName().toLowerCase() == plugin) {
                                    it.onUnload(bot)

                                    message.setText("Plugin unloaded!")
                                    connection.send(message.build())
                                }
                            }
                        }
                    }
                } else {
                    when (args[0]) {
                        "list" -> {
                            bot.plugins.forEach {
                                message.setText(it.getName())
                                connection.send(message.build())
                            }
                        }
                    }
                }
            }
        })

        PluginHelper.instance.addHelp(PluginHelp(
                ".plugin list",
                IRCBotAccess.Level.ROOT,
                "List all available plugins"))

        PluginHelper.instance.addHelp(PluginHelp(
                ".plugin load",
                IRCBotAccess.Level.ROOT,
                "Load a plugin",
                "plugin"))

        PluginHelper.instance.addHelp(PluginHelp(
                ".plugin unload",
                IRCBotAccess.Level.ROOT,
                "Unload a plugin",
                "plugin"))
    }

    override fun onUnload(bot: IRCBot) {
        bot.removeListener(".plugin")
        arrayOf("list", "load", "unload").forEach {
            PluginHelper.instance.removeHelp(it)
        }
    }
}